pkg/labels/labels.go

```diff
@@ -267,9 +267,45 @@ func (l *Label) IsReservedSource() bool {
 	return l.Source == LabelSourceReserved
 }
 
-// matches returns true if l matches the target
-func (l *Label) matches(target *Label) bool {
-	return l.Equals(target)
+// Matches returns true if l Matches the target.
+// note that this is a loose match:
+// - l.Source == any means sources must not match
+// - If both are CIDR labels, and l contains target, it will match.
+func (l *Label) Matches(target *Label) bool {
+	return l.MatchesKey(target) && l.Value == target.Value
+}
+
+// MatchesKey returns true if l has target's key.
+// note that l may be looser than target.
+func (l *Label) MatchesKey(target *Label) bool {
+	if !l.IsAnySource() && l.Source != target.Source {
+		return false
+	}
+
+	// Do cidr-aware matching if both sources are "cidr".
+	// This means that "cidr:10.0.0.0/24".matchesKey("cidr:10.0.0.1/32") = true
+	// Note that this does not take effect if either label has an "any" key.
```

Comment:
I don't quite understand the label to label level here but I would think that a selector for `"any:10.0.0.0/24".matchesKey("cidr:10.0.0.1/32") == true`. Source any at the selector level is ambivalent about the source of the actual label it's targeting. Little hard to see through the abstractions to here given this is on a label-by-label basis but this surprises me a bit.

Reply:
I'm concerned about accidentally matching CIDRs for the `any:` case. CIDR parsing is somewhat lenient; I want to play around and see if there are any textual labels that might trigger an IPv6 prefix. That would be a rude surprise.

Reply:
Is this only restricting source checking for CIDR labels or for all labels? There is a small chance that someone, somewhere, relies on the current `any` behavior. Probably not for CIDRs specifically - I'd expect that the policy parsing for ToCIDR statements is sanitizing this already. But maybe for other labels.

Reply:
Just cidr behavior.

FWIW, “any” selection is used extensively already by the policy engine. So I had broken that here, all the tests would be broken.